modules/javafx.graphics/src/main/native-glass/gtk/glass_window_ime.cpp

```diff
@@ -23,266 +23,159 @@
-    }
+    jdoubleArray pos = (jdoubleArray)mainEnv->CallObjectMethod(get_jview(),
+                                      jViewNotifyInputMethodCandidateRelativePosRequest,
+                                      0);
 
-    for (i = 0; i < styles->count_styles; ++i) {
-        if (styles->supported_styles[i] == (XIMPreeditCallbacks | XIMStatusNothing)
-                || styles->supported_styles[i] == (XIMPreeditNothing | XIMStatusNothing)) {
-            result = styles->supported_styles[i];
-            break;
-        }
-    }
+    nativePos = mainEnv->GetDoubleArrayElements(pos, NULL);
 
-    XFree(styles);
+    GdkRectangle rect;
+    if (nativePos) {
+        rect.x = (int) nativePos[0];
+        rect.y = (int) nativePos[1];
+        rect.width = 0;
+        rect.height = 0;
 
-    return result;
+        mainEnv->ReleaseDoubleArrayElements(pos, nativePos, 0);
+        gtk_im_context_set_cursor_location(im_ctx.ctx, &rect);
+    }
 }
 
 void WindowContextBase::enableOrResetIME() {
-    Display *display = gdk_x11_display_get_xdisplay(gdk_window_get_display(gdk_window));
-    if (xim.im == NULL || xim.ic == NULL) {
-        xim.im = XOpenIM(display, NULL, NULL, NULL);
-        if (xim.im == NULL) {
-            return;
-        }
-
-        XIMStyle styles = get_best_supported_style(xim.im);
-        if (styles == 0) {
-            return;
-        }
-
-        XIMCallback startCallback = {(XPointer) jview, (XIMProc) (void *) im_preedit_start};
-        XIMCallback doneCallback = {(XPointer) jview, im_preedit_done};
-        XIMCallback drawCallback = {(XPointer) jview, im_preedit_draw};
-        XIMCallback caretCallback = {(XPointer) jview, im_preedit_caret};
-
-        XVaNestedList list = XVaCreateNestedList(0,
-                XNPreeditStartCallback, &startCallback,
-                XNPreeditDoneCallback, &doneCallback,
-                XNPreeditDrawCallback, &drawCallback,
-                XNPreeditCaretCallback, &caretCallback,
-                NULL);
-
-        xim.ic = XCreateIC(xim.im,
-                XNInputStyle, styles,
-                XNClientWindow, GDK_WINDOW_XID(gdk_window),
-                XNPreeditAttributes, list,
-                NULL);
-
-        XFree(list);
-
-        if (xim.ic == NULL) {
-            return;
-        }
+    if (im_ctx.on_preedit) {
+        gtk_im_context_focus_out(im_ctx.ctx);
     }
 
-    if (xim.enabled) { //called when changed focus to different input
-        XmbResetIC(xim.ic);
+    if (!im_ctx.enabled) {
+        im_ctx.ctx = gtk_im_multicontext_new();
+        gtk_im_context_set_client_window(GTK_IM_CONTEXT(im_ctx.ctx), gdk_window);
+        g_signal_connect(im_ctx.ctx, "preedit-start", G_CALLBACK(on_preedit_start), this);
+        g_signal_connect(im_ctx.ctx, "preedit-changed", G_CALLBACK(on_preedit_changed), this);
+        g_signal_connect(im_ctx.ctx, "preedit-end", G_CALLBACK(on_preedit_end), this);
+        g_signal_connect(im_ctx.ctx, "commit", G_CALLBACK(on_commit), this);
```

Comment:
On Ubuntu 22.04 using the Japanese (Mozc) input method I get a bunch of debug output from IBus. The message reads

`IBUS-WARNING **: 08:50:12.994: java has no capability of surrounding-text feature`

I was able to silence this by connecting to the surrounding text signal and simply returning TRUE from the callback.

Reply:
Fixed, although JavaFX does not support it, so the warning is true.